activerecord/lib/active_record/connection_adapters/abstract/connection_pool.rb

```diff
@@ -370,8 +370,8 @@ def lock_thread=(lock_thread)
       #
       # #connection can be called any number of times; the connection is
       # held in a cache keyed by a thread.
-      def connection
-        @thread_cached_conns[connection_cache_key(@lock_thread || Thread.current)] ||= checkout
+      def connection(owner_thread = @lock_thread || Thread.current)
```

Comment:
Yeah, I'd like some other ideas on how to proceed without affecting the public API if possible. As [noted above](https://github.com/rails/rails/pull/28207#issuecomment-283217928), the underlying issue is that some of the existing [connection pool tests](https://github.com/rails/rails/pull/28207/files#diff-4dec8806a210fe9ff8021207675d4f7c) (likely some end-user tests as well) depend upon the original behavior where threads are not locked to a single connection, so in order to get the tests to pass with minimal changes I chose to modify the public API to allow the new behavior to be optionally overridden. Part of my rationale for choosing this approach was that `#release_connection` already contained `owner_thread` in its public API, so this would just update `#connection` and `#with_connection` to match.

If we don't modify `#connection` and `#with_connection` to match the existing `#release_connection` signature, then we'll need some other way of allowing the lock-thread behavior to be overridden for tests that break on it. Perhaps a switch at the test-case-class or test-method level to mark specific tests to either run outside of a test-transaction entirely, and/or disable the transaction's thread-locking behavior for that test/test-case?